### PR TITLE
Throwing exception when passing empty list of values as condition

### LINKED ISSUE #7092: ORM: Unexpected behavior with 'WHERE NOT IN' and empty array
```php
$articles->find()->where(function($exp,$q){
    return $exp->notIn('id', []);
});

# WHERE 1 != 1
```
So this query returns no rows.
This may be a feature, but this is the opposite of what I expected.

In CakePHP 2.x:
```php
$Article->find('all', [
	'conditions' => [
		'id !=' => [],
	]
]);

# WHERE `id` IS NOT NULL
```

In http://www.doctrine-project.org/jira/browse/DDC-3063 , they are talking about a same issue.

### PR DESCRIPTION
**Pull request by lorenzo**
Fixes #7092

### CODE REVIEW

**Comment by AD7six** [comment]
:+1:

File: src/Database/Expression/Comparison.php [hunk 1]
```diff
@@ -189,9 +190,12 @@ protected function _stringExpression($generator)
             $value = $this->_flattenValue($this->_value, $generator, $type);
 
             // To avoid SQL errors when comparing a field to a list of empty values,
-            // generate a condition that will always evaluate to false
+            // better just throw an exception here
             if ($value === '') {
-                return ['1 != 1', ''];
+                $field = $this->_field instanceof ExpressionInterface ? $this->_field->sql($generator) : $this->_field;
+                throw new DatabaseException(
+                    "Imposible to generate condition with empty list of values for field ($field)"
```
Comment (HavokInspiration) [inline src/Database/Expression/Comparison.php]: Impossible*

**Comment by dereuromark** [comment]
Better than silently building 1!=1 :+1: 
But sometimes I really miss the good old auto-IN adding as convenience logic.

File: src/Database/Expression/Comparison.php [hunk 2]
```diff
@@ -189,9 +190,12 @@ protected function _stringExpression($generator)
             $value = $this->_flattenValue($this->_value, $generator, $type);
 
             // To avoid SQL errors when comparing a field to a list of empty values,
-            // generate a condition that will always evaluate to false
+            // better just throw an exception here
             if ($value === '') {
-                return ['1 != 1', ''];
+                $field = $this->_field instanceof ExpressionInterface ? $this->_field->sql($generator) : $this->_field;
+                throw new DatabaseException(
+                    "Impossible to generate condition with empty list of values for field ($field)"
```
Comment (ADmad) [inline src/Database/Expression/Comparison.php]: Can we use `sprintf` instead of variable inside string :smile:

Comment (lorenzo) [inline src/Database/Expression/Comparison.php]: made the line even longer, that's why I did this :P

Comment (ADmad) [inline src/Database/Expression/Comparison.php]:
```php
throw new DatabaseException(sprintf(
    'Impossible to generate condition with empty list of values for field (%s)',
    $field
));
```
:wink:

Comment (dereuromark) [inline src/Database/Expression/Comparison.php]: @ADmad Technically the two inner lines would need an extra level of indentation then, though :)

Comment (ADmad) [inline src/Database/Expression/Comparison.php]: PSR2 doesn't complain so "techinially" there is no problem.

Comment (dereuromark) [inline src/Database/Expression/Comparison.php]: Just because it doesnt complain doesnt make it correct ;) I disagree with that logic.

Comment (ADmad) [inline src/Database/Expression/Comparison.php]: Then let's agree to disagree.

Comment (dereuromark) [inline src/Database/Expression/Comparison.php]: PS: It also doesnt complain about not using sprintf() - so @lorenzo can kindly ignore your comment then :)

**Comment by markstory** [comment]
Nice work @lorenzo

**Comment by chinpei215** [comment]
:+1:

**Pull request merged**